### PR TITLE
IGNITE-18087 Populate DistributionZoneManager with CMG listeners to logical topology events

### PR DESCRIPTION
**Pull request by alievmirza**
https://issues.apache.org/jira/browse/IGNITE-18087

### CODE REVIEW

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 1]
```diff
@@ -373,4 +404,92 @@ private void updateMetaStorageOnZoneDelete(int zoneId, long revision) {
...
+    private void initMetaStorageKeysOnStart() {
+        logicalTopologyService.logicalTopologyOnLeader().thenAccept(snapshot -> {
+            long topologyVersionFromCmg = snapshot.version();
+
+            byte[] topVerFromMetastorage;
+
+            try {
+                topVerFromMetastorage = metaStorageManager.get(zonesLogicalTopologyVersionKey()).get().value();
+            } catch (InterruptedException | ExecutionException e) {
+                throw new RuntimeException(e);
```
Comment (sergeyuttsel) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:464]: Need to use IgniteInternalException or other.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java [hunk 2]
```diff
@@ -36,19 +37,48 @@ class DistributionZonesUtil {
     /** Key prefix for zone's data nodes. */
     private static final String DISTRIBUTION_ZONE_DATA_NODES_PREFIX = "distributionZone.dataNodes.";
 
+    /** Key prefix for zones' logical topology nodes. */
+    private static final String DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_PREFIX = "distributionZones.logicalTopology";
```
Comment (sergeyuttsel) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java:41]: it isn't a prefix. It's just a DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY or DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_KEY

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java [hunk 3]
```diff
@@ -36,19 +37,48 @@ class DistributionZonesUtil {
...
+
+    /** Key prefix for zones' logical topology version. */
+    private static final String DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_VERSION_PREFIX = "distributionZones.logicalTopologyVersion";
```
Comment (sergeyuttsel) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java:44]: Not a prefix.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java [hunk 4]
```diff
@@ -36,19 +37,48 @@ class DistributionZonesUtil {
...
     /** ByteArray representation of {@link DistributionZonesUtil#DISTRIBUTION_ZONE_DATA_NODES_PREFIX}. */
     static ByteArray zoneDataNodesKey(int zoneId) {
         return new ByteArray(DISTRIBUTION_ZONE_DATA_NODES_PREFIX + zoneId);
     }
 
+    /** ByteArray representation of {@link DistributionZonesUtil#DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_PREFIX}. */
+    private static final ByteArray DISTRIBUTION_ZONE_LOGICAL_TOPOLOGY_KEY = new ByteArray(DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_PREFIX);
+
+    /** ByteArray representation of {@link DistributionZonesUtil#DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_VERSION_PREFIX}. */
+    private static final ByteArray DISTRIBUTION_ZONES_LOGICAL_TOPOLOGY_VERSION_KEY =
```
Comment (sergeyuttsel) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZonesUtil.java:58]: Need to move "private static final" fields on top.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 5]
```diff
@@ -74,21 +82,45 @@ public class DistributionZoneManager implements IgniteComponent {
...
     private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
 
+    /** Logical topology service to track topology changes. */
+    private final LogicalTopologyService logicalTopologyService;
+
+    /** Listener for a topology events. */
+    private final LogicalTopologyEventListener topologyEventListener = new LogicalTopologyEventListener() {
+        @Override
+        public void onAppeared(ClusterNode appearedNode, LogicalTopologySnapshot newTopology) {
+            updateMetaStorageKeys(newTopology, false);
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:92]: Confusing naming, I'd rather use updateLogicalTopologyInMetaStorage.

**Review by sanpwc — CHANGES_REQUESTED** [review]

**Review by sanpwc — APPROVED** [review]

**Pull request merged**